### PR TITLE
Add flag to disallow decorating typed functions with untyped decorators

### PR DESCRIPTION
**Pull request by ilinum**
If none of the argument types + return type of the decorated function are Any and the decorator has return type `Any`, `--disallow-any=decorator` reports an error.

### CODE REVIEW

**Comment by ilinum** [comment]
Note that this does not address decorators that return `Callable[..., X]`.

**Comment by ddfisher** [comment]
It seems to me that catching decorators that return `Callable[..., X]` is important -- can support for that be added to this PR?

**Comment by ilinum** [comment]
Yeah, I'll work on that!

**Comment by ilinum** [comment]
With the most recent commit, this PR now addresses decorators that return `Callable[..., X]`.
Basically, it reports an error for every parameter that became `Any` because of a decorator.

File: test-data/unit/check-flags.test [hunk 1]
```diff
@@ -515,3 +515,108 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
+[case testDisallowAnyDecoratorSimple]
+# flags: --disallow-any=decorator
+from typing import Any
+
+def d(f):
+    return f
+
+@d  # no error
+def f(x: Any) -> Any: pass
+@d  # error
```
Comment (ilinum) [inline test-data/unit/check-flags.test]: Woah, I had no idea you can have multiple `# E: ...` on the same line!

File: mypy/messages.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 import re
 import difflib
 
-from typing import cast, List, Dict, Any, Sequence, Iterable, Tuple
+from typing import cast, List, Dict, Any, Sequence, Iterable, Tuple, Optional
```
Comment (ddfisher) [inline mypy/messages.py]: Unused import.

File: test-data/unit/check-flags.test [hunk 3]
```diff
@@ -515,3 +515,95 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
+@d
+def f() -> None: pass  # E: Function is untyped after decorator transformation
+[builtins fixtures/list.pyi]
+
+[case testDisallowAnyDecoratedPartlyTypedCallable]
+# flags: --disallow-any=decorated --ignore-missing-imports
+from typing import Callable, Any, List
+
+def d(f) -> Callable[[int, Any], Any]: pass
+def d2(f) -> Callable[[int], List[Any]]: pass
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: It'd be good to also test a callable which doesn't have Any as part of its return value.

File: test-data/unit/check-flags.test [hunk 4]
```diff
@@ -515,3 +515,95 @@ x, y = 1, 2  # type: Unchecked, Unchecked
 [out]
 main:4: error: Type of variable becomes "Any" due to an unfollowed import
 main:6: error: A type on this line becomes "Any" due to an unfollowed import
+
+[case testDisallowAnyDecoratedUnannotatedDecorator]
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: Some additional tests I'd like to see:
- test that unannotated functions and functions annotated with `Any` that are undecorated do not produce an error
- test the behavior on a function that is has two decorators -- one which makes it untyped and then one which makes it typed (without Anys) again

File: mypy/messages.py (hunk 2 above)
Comment (ilinum) [inline mypy/messages.py]: Whoops!
[Fixed](https://github.com/python/mypy/pull/3536/commits/2eb7b61d0a1d6eb142a55be16f827fae5939df42).

File: test-data/unit/check-flags.test (hunk 3 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: [Done](https://github.com/python/mypy/pull/3536/commits/3b7a079d189cf82da3318d20b99eb8d7e0a68b7e)!

File: test-data/unit/check-flags.test (hunk 4 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: Huh, both are really cool test ideas, actually, especially the second one.

[Done](https://github.com/python/mypy/pull/3536/commits/3b7a079d189cf82da3318d20b99eb8d7e0a68b7e)!

File: test-data/unit/check-flags.test (hunk 3 above)
Comment (ddfisher) [inline test-data/unit/check-flags.test]: Sorry, I was unclear.  The test you added is good, but it'd also be worth testing a Callable that contains Any but _not_ as part of its return value.  (It looks to me like this case is currently not covered by any of the tests.)

Comment (ilinum) [inline test-data/unit/check-flags.test]: All right, this should be good now.
Test with Callable that takes parameters of type `Any` but returns not `Any`.

**Pull request merged**